### PR TITLE
Migrated helper menu to a svelte object

### PR DESCRIPTION
**Pull request by ahanden**
re #540

### CODE REVIEW

**Comment by atompkins** [comment]
There are some good ideas here. It's almost a straight conversion.
I'm concerned about mounting the component on every page load.
Previously we only rendered it on mouseenter.
On my M1 with 6x CPU throttle it only takes 6ms.
It's possible that the new version of svelte addresses some of the
performance issues. I just haven't had time recently to properly
look into it. I'll setup an environment on a Pi4 to see.

Rather than a straight conversion I would like to use Modal.svelte.
I was hoping to use some of the behaviour like click anywhere to
close.
Your version doesn't close the menu when something gets
clicked. This was something I added "recently" #593. I don't like
it hanging around until you click on it to close it.

I like functionLookup, it's nice and explicit. I don't think you need
'type' though. An item either has a function or a href.
I don't understand why dev quick links isn't part of
functionLookup though.
I don't understand why the hasOwnProperty check is needed?

sendMsg event is missing and I'd like events on href as well.

**Comment by ahanden** [comment]
Yeah, the `hasOwnProperty ` bit was from my brain crossing wires. I've been coding a lot in python again.

**Comment by atompkins** [comment]
on the Pi4 unthrottled, injectHelperMenu takes 52ms.
for context injectQuickLinks takes 34ms.
i'm thinking we can live with that.
just bear it in mind if we add any more.

File: src/modules/chrome/helperMenu/injectHelperMenu.js [hunk 1]
```diff
@@ -1,91 +1,14 @@
...
-    helperMenu.classList.add('helperMenuFixed');
-  }
-  once(helperMenu, 'mouseenter', showHelperMenu);
-  if (getValue('draggableHelperMenu')) {
-    helperMenu.classList.add('helperMenuMove');
-    draggable(helperMenu);
-  }
-  insertElement(document.body, helperMenu);
+  const target = querySelector('.mainbody');
+  startApp({}, target);
```
Comment (atompkins) [inline src/modules/chrome/helperMenu/injectHelperMenu.js:13]: needs guard

File: src/modules/chrome/helperMenu/functionLookup.js [hunk 2]
```diff
@@ -19,25 +21,118 @@ import setmgr from '../pageSwitcher/loader/setmgr';
...
+        label: 'Creature Log', fn: creatureLog,
+      },
+      {
+        label: 'Recipe Manager', fn: recipeMgr,
+      },
+      {
+        label: 'Quick Links', fn: quickLinksManager,
+      },
+      {
+        label: 'Inventory Manager', hrefn: `${notepadBlankUrl}invmanagernew`,
```
Comment (atompkins) [inline src/modules/chrome/helperMenu/functionLookup.js:48]: hrefn looks like a typo or a bad merge?

File: src/modules/chrome/helperMenu/HelperMenu.svelte [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+</script>
+<button
+  class:helperMenuFixed={ isFixed }
+  class:helperMenuMove={ isDraggable }
+  type="button"
+  id="toggle"
+  on:click={ toggle }>
+  Helper Menu
+</button>
+<Modal visible={ showMenu } on:close={ toggle }>
```
Comment (atompkins) [inline src/modules/chrome/helperMenu/HelperMenu.svelte:44]: is there a way to make the modal appear underneath the HelperMenu toggle button?

File: src/modules/chrome/helperMenu/HelperMenu.svelte [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+          { :else if menuItem.href }
+            <a href={ menuItem.href } on:click={ () => sendHelperEvent(menuItem.label) }>
+              { menuItem.label }
+            </a>
+          { :else if menuItem.playerName }
+            <button
+              type="button"
+              class="helperDevBtn"
+              on:click={ () => { toggle(); sendHelperEvent('sendMsg'); } }
+              onclick="openQuickMsgDialog('{ menuItem.playerName }');">
```
Comment (atompkins) [inline src/modules/chrome/helperMenu/HelperMenu.svelte:63]: i particularly don't like inline js getting eval'ed. onclick shows up in the DOM whereas on:click doesn't.
there are 3 lines of code being executed on:click. i think this should be in a function.
alternatively you can have multiple on:click, might be ugly though.

**Review by atompkins — CHANGES_REQUESTED** [review]
Apart from a couple of minor niggles the main issue here is positioning. It was surprising to see the menu in the centre of the page. This is most likely the reason I hadn't tackled it 😝 
Also in terms of ergonomics, when you select a modal function, the menu and the background disappear and then the new modal appears with the background again. I think it would be better if the new modal stacked on top before the menu disappears. Less jarring?

**Comment by ahanden** [comment]
>  Also in terms of ergonomics, when you select a modal function, the menu and the background disappear and then the new modal appears with the background again. I think it would be better if the new modal stacked on top before the menu disappears. Less jarring?

I'm stumped on how to do this one.

**Comment by atompkins** [comment]
1. functionLookup contains "Buff Log" twice. The one in "Guild" looks like a copy paste error?
2. For the open/close thing i'd play around with awaiting the loader. i tried returning the runDefault from a few of the loaders and it seemed ok. i'm not entirely convinced it's correct though.
3. there is a weird positioning problem if you don't move "HelperMenu"
image
image
i'm really not convinced about this approach but i don't have time to experiment right now. i was expecting to make the modal relative to the "HelperMenu" button.

**Comment by ahanden** [comment]
I ended up removing the Modal.svelte again. I know the "click anywhere to close" feature may be nice, but I really don't think the benefit is worth the cost.

Here the menu moves with the button (javascript free), there's no complicated queuing between svelte modules, and the user can still see their page while the menu is open.

**Pull request merged**